x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java

```diff
@@ -174,34 +174,38 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
 
     @Override
     protected Query doToQuery(SearchExecutionContext context) throws IOException {
-        if ((pinnedIds != null && pinnedIds.isEmpty() == false) && (pinnedDocs != null && pinnedDocs.isEmpty() == false)) {
-            throw new IllegalArgumentException("applied rules contain both pinned ids and pinned docs, only one of ids or docs is allowed");
-        }
-
-        if (pinnedIds != null && pinnedIds.isEmpty() == false) {
-            PinnedQueryBuilder pinnedQueryBuilder = new PinnedQueryBuilder(organicQuery, pinnedIds.toArray(new String[0]));
-            return pinnedQueryBuilder.toQuery(context);
-        } else if (pinnedDocs != null && pinnedDocs.isEmpty() == false) {
-            PinnedQueryBuilder pinnedQueryBuilder = new PinnedQueryBuilder(organicQuery, pinnedDocs.toArray(new Item[0]));
-            return pinnedQueryBuilder.toQuery(context);
-        } else {
-            return organicQuery.toQuery(context);
-        }
-
+        throw new IllegalStateException(NAME + " should have been rewritten to another query type");
```

Comment:
Reading through the change again this part might be considered breaking. Coordinator nodes in previous version will continue to send `RuleQueryBuilder` to shards in newer versions. That's probably acceptable in technical preview but wanted to call out.

Reply:
Good question, I'd say yes to ensure that users are not surprised by this even though it's in technical preview. The other option is to try to avoid the breaking change by allowing `doToQuery` to work with queries coming from older nodes?

Reply:
I added a TransportVersion check here. The Features API is not easily accessible from the QueryBuilder that I could see.